### PR TITLE
Make files public and private

### PR DESCRIPTION
**Pull request by ryanseys**
Included in this PR is my interpretation of how we should handle making Google Cloud Storage objects public and private.

I am not putting it under the `file.acl` object because if the user understand acls then they probably don't need to use these helper methods. These are slap-em-on permissions that require minimal configuration.

Buckets, by default, apply a default ACL of `project-private`, that is it is private only to the maintainers of the project. This is the "default" setting. By running `file.makePrivate()` we are reverting to this default. If you want to REALLY make it private, that is only to the object owner user, then you can pass in `{ strict: true }` (ideas on a better param name are welcome). `makePrivate()` removes all other permissions. It DNGAF.

`makePublic()` just slaps on `allUsers: READER` permission, maintaining all other permissions. It's nicer than `makePrivate()` in this regard.

Edit: Added tests.

### CODE REVIEW

**Comment by stephenplusplus** [comment]
All sounds good to me.

**Comment by ryanseys** [comment]
Cool beans, I'll add some tests to this bad boy. Might leave makePublic and makePrivate for buckets out for now as a way to break the problem down.

File: test/storage/file.js [hunk 1]
```diff
@@ -1025,6 +1025,22 @@ describe('File', function() {
     });
   });
 
+  describe('makePublic', function() {
+    it('should make the file public', function(done) {
+
+      // If this mocked method (THAT DOES NOTHING!) is removed, nothing works.
+      file.acl.makeAclObject_ = function() {};
```
Comment (ryanseys) [inline test/storage/file.js]: @stephenplusplus Maybe you can explain this. If I take this line out, this test does not pass. So much is being mocked, it's hard to follow where things are getting overwritten and why this is required or what it's doing!

Comment (stephenplusplus) [inline test/storage/file.js]: Quick answer: You should be testing that when you call `file.makePublic` it calls `this.acl.add` with the correct arguments.

Philosophical explanation: It's not the business of this test to be concerned with how `this.acl.add` does its job - we have tests for that. That's why you have to mock `file.acl.makeAclObject_`, because `this.acl.add` is using that method to do its job. And again, our test of `makePublic` only cares that `this.acl.add` is getting what it needs, and if we trust our acl test suite, then we don't need to question it from way up here.

Comment (ryanseys) [inline test/storage/file.js]: Is this a side-effect of using mockery? I had a feeling I should have been doing that instead. I'll update it to do that.

Comment (stephenplusplus) [inline test/storage/file.js]: This is just how I see effective unit testing working for our library, not specific to this situation. `file.acl.makeAclObject_` is being called because when you call `file.makePublic`, it in turn calls `file.acl.add`, which calls `file.acl.makeAclObject_`. Instead of letting it fall through all those places, we should interrupt it on its way to `file.acl.add`.

Comment (ryanseys) [inline test/storage/file.js]: That makes sense. Just doesn't make sense that I added an empty function and it still worked.

**Comment by ryanseys** [comment]
This should be good for a second look. I've added tests. I've omitted a test for testing makePrivate with `strict: true` because I only have project-level access, not user access, so the file would be completely inaccessible to me after making it private. I do test makePrivate on a project level though.

File: lib/storage/file.js [hunk 2]
```diff
@@ -798,6 +798,68 @@ File.prototype.setMetadata = function(metadata, callback) {
...
+ * file.makePrivate({ strict: true }, function(err) {});
+ */
+File.prototype.makePrivate = function(options, callback) {
+  var that = this;
+  if(util.is(options, 'function')) {
+      callback = options;
+      options = {};
+  }
+  var path = '/o/' + encodeURIComponent(this.name);
+  var query = { predefinedAcl: (options.strict ? '' : 'project') + 'private' };
```
Comment (stephenplusplus) [inline lib/storage/file.js]: In [the docs](https://cloud.google.com/storage/docs/json_api/v1/objects/patch), it asks for `projectPrivate` - does the casing matter?

File: lib/storage/file.js [hunk 3]
```diff
@@ -798,6 +798,68 @@ File.prototype.setMetadata = function(metadata, callback) {
...
+ * file.makePrivate(function(err) {});
+ *
+ * //-
+ * // Set the file private so only owner can see and modify it.
+ * //-
+ * file.makePrivate({ strict: true }, function(err) {});
+ */
+File.prototype.makePrivate = function(options, callback) {
+  var that = this;
+  if(util.is(options, 'function')) {
```
Comment (stephenplusplus) [inline lib/storage/file.js]: `if (`

File: lib/storage/file.js [hunk 4]
```diff
@@ -798,6 +798,68 @@ File.prototype.setMetadata = function(metadata, callback) {
...
+      callback = options;
+      options = {};
+  }
+  var path = '/o/' + encodeURIComponent(this.name);
+  var query = { predefinedAcl: (options.strict ? '' : 'project') + 'private' };
+  var metadata = { acl: null };
```
Comment (stephenplusplus) [inline lib/storage/file.js]: Can you leave a quick comment in the code explaining why we're doing this?

File: lib/storage/file.js [hunk 5]
```diff
@@ -798,6 +798,68 @@ File.prototype.setMetadata = function(metadata, callback) {
...
+
+/**
+ * Set a file to be publicly readable and maintains all previous permissions.
+ *
+ * @param {function=} callback - The callback function.
+ *
+ * @example
+ * file.makePublic(function(err) {});
+ */
+File.prototype.makePublic = function(callback) {
```
Comment (stephenplusplus) [inline lib/storage/file.js]: We need a `callback = callback || util.noop`

File: lib/storage/file.js [hunk 6]
```diff
@@ -798,6 +798,68 @@ File.prototype.setMetadata = function(metadata, callback) {
 };
 
 /**
+ * Set a file to be private to the project and removes all other permissions.
+ * If you set `options.strict` to true, it will set it to be private to the
+ * owner user only.
```
Comment (stephenplusplus) [inline lib/storage/file.js]: Hope you don't mind these small change suggestions:

```diff
- Set a file to be private to the project and removes all other permissions.
+ Make a file private to the project and remove all other permissions.

- If you set `options.strict` to true, it will set it to be private to the owner user only.
+ Set `options.strict` to true to make the file private to only the owner.
```

File: lib/storage/file.js [hunk 7]
```diff
@@ -798,6 +798,68 @@ File.prototype.setMetadata = function(metadata, callback) {
...
+ *
+ * @example
+ *
+ * //-
+ * // Set the file private so only project maintainers can see and modify it.
+ * //-
+ * file.makePrivate(function(err) {});
+ *
+ * //-
+ * // Set the file private so only owner can see and modify it.
```
Comment (stephenplusplus) [inline lib/storage/file.js]: `only the owner`

File: lib/storage/file.js [hunk 8]
```diff
@@ -798,6 +798,68 @@ File.prototype.setMetadata = function(metadata, callback) {
...
+    }
+
+    that.metadata = resp;
+
+    callback(null);
+  });
+};
+
+/**
+ * Set a file to be publicly readable and maintains all previous permissions.
```
Comment (stephenplusplus) [inline lib/storage/file.js]: `maintain`

File: lib/storage/file.js (hunk 2 above)
Comment (ryanseys) [inline lib/storage/file.js]: Nope doesn't matter, I'll update though for clarity.

File: lib/storage/file.js (hunk 5 above)
Comment (ryanseys) [inline lib/storage/file.js]: Done.

**Comment by ryanseys** [comment]
PTAL.

File: lib/storage/file.js [hunk 9]
```diff
@@ -798,6 +798,73 @@ File.prototype.setMetadata = function(metadata, callback) {
...
+ * //-
+ * file.makePrivate(function(err) {});
+ *
+ * //-
+ * // Set the file private so only the owner can see and modify it.
+ * //-
+ * file.makePrivate({ strict: true }, function(err) {});
+ */
+File.prototype.makePrivate = function(options, callback) {
+  var that = this;
```
Comment (stephenplusplus) [inline lib/storage/file.js]: We're coming for you, `that`!!!

**Comment by stephenplusplus** [comment]
LGTM but I'll leave it unmerged for now to confirm the way it works is what @jgeewax had in mind.

**Comment by stephenplusplus** [comment]
Yolo merge?

**Comment by ryanseys** [comment]
YEP

**Comment by stephenplusplus** [comment]
Woooooo!

**Comment by jgeewax** [comment]
can i just say for a minute how great this thread is? @ryanseys @stephenplusplus

**Comment by ryanseys** [comment]
Glad you like! :+1: :dancer:
